### PR TITLE
Example of configuring builtin plugin.

### PR DESCRIPTION
**Pull request by monopole**
Requires #1532

### CODE REVIEW

**Comment by monopole** [comment]
@richardmarshall  @jbrette PTAL

File: examples/configureBuiltinPlugin.md [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+ | `secretGenerator` | _SecretGenerator_ |
+ | `configMapGenerator` | _ConfigMapGenerator_ |
+ | `namespace` | _NamespaceTransformer_ |
+ | `patches`   | _PatchTransformer_ |
+ | `commonLabels` | _LabelTransformer_ |
+ | `commonAnnotations` | _AnnotationTransformer_ |
+ | `namePrefix`  | _PrefixSuffixTransformer_ |
+ | `nameSuffix`  | _PrefixSuffixTransformer_ |
+ | `image` | _ImageTagTransformer_ |
+ | `replicas` | _ReplicasTransformer_ |
```
Comment (Liujingfang1) [inline examples/configureBuiltinPlugin.md]: How about adding `patchesStrategicMerge` and `patchesJson6902` as well since we haven't deprecated them?

File: examples/configureBuiltinPlugin.md [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+```
+
+Then make the two plugin configuration files
+referred to by the trans
+
+<!-- @makeAnnotatorPluginConfig @testAgainstLatestRelease -->
+```
+cat <<EOF >$DEMO_HOME/myAnnotator.yaml
+apiVersion: builtin
+kind: AnnotationsTransformer
```
Comment (Liujingfang1) [inline examples/configureBuiltinPlugin.md]: Can we document what fields can be used in each type of builtin transformer/generator YAML? Maybe list a few YAML files for them at the end of this example.

File: docs/plugins/builtins.md [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+> types.SecretArgs
+> ```
+
+#### Example
+
+> ```
+> apiVersion: builtin
+> kind: SecretGenerator
+> metadata:
+>   name: mySecret
```
Comment (jbrette) [inline docs/plugins/builtins.md]: this name is not DNS compliant, and is rejected by kubectl

File: docs/plugins/builtins.md [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+key/value as data.
+
+Each configMapGenerator item accepts a parameter of
+`behavior: [create|replace|merge]`.
+This allows an overlay to modify or
+replace an existing configMap from the parent.
+
+```
+configMapGenerator:
+- name: myJavaServerProps
```
Comment (jbrette) [inline docs/plugins/builtins.md]: this name is not DNS compliant.

File: docs/plugins/builtins.md [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+> types.GeneratorOptions
+> types.ConfigMapArgs
+> ```
+
+#### Example
+> ```
+> apiVersion: builtin
+> kind: ConfigMapGenerator
+> metadata:
+>   name: myMap
```
Comment (jbrette) [inline docs/plugins/builtins.md]: this name is not DNS compliant.

File: docs/plugins/builtins.md [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+> ImageTag   image.Image
+> FieldSpecs []config.FieldSpec
+> ```
+
+#### Example
+> ```
+> apiVersion: builtin
+> kind: ImageTagTransformer
+> metadata:
+>   name: notImportantHere
```
Comment (jbrette) [inline docs/plugins/builtins.md]: This is actually really important. In order to be able to change two images, you need to add multiple transformers. The name of the transformers have to be different, otherwise the resources accumulation will fail

File: docs/plugins/builtins.md [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+> FieldSpecs []config.FieldSpec
+> ```
+
+#### Example
+> ```
+> apiVersion: builtin
+>  kind: NamespaceTransformer
+>  metadata:
+>    name: notImportantHere
+>    namespace: test
```
Comment (jbrette) [inline docs/plugins/builtins.md]: The syntax of the namespace transformer is kind of misleading. the "metadata/name" is not important but the "metadata/namespace" is. Unlike the other transformer (prefix, suffix), that field is not part of the "spec" of the Transformer but part of the metadata.

Comment (Liujingfang1) [reply docs/plugins/builtins.md]: Not every plugin has the config under `spec`. Even for prefix/suffix transformer, it's `prefix` and `suffix` instead of `spec`. The transformer/generator plugin read the whole k8s styled yaml as its configuration, not only `spec` field.

Comment (jbrette) [reply docs/plugins/builtins.md]: Let me rephrase. All the other transformers have the addditional parameters at the top level....basically the fieldSpecs level, the metadata field is actually irrelevant most of the time.

```yaml
apiVersion: builtin
kind: LabelTransformer
metadata:
  name: notImportantHere
labels:
  app: myApp
  env: production
fieldSpecs:
  - path: metadata/labels
    create: true
```

or 

```yaml
apiVersion: builtin
kind: PrefixSuffixTransformer
metadata:
  name: notImportantHere
prefix: baked-
suffix: -pie
fieldSpecs:
  - path: metadata/name
```

But not for the namespace transformer. The consistent way of doing, should be like this:

```yaml
apiVersion: builtin
kind: NamespaceTransformer
metadata:
  name: notImportantHere
namespace: test
fieldSpecs:
  - path: metadata/namespace
```

The other issue is [here](https://github.com/kubernetes-sigs/kustomize/blob/master/pkg/target/kusttarget.go#L313). The name is actually important for ImageTransformer if you need multiple image transformations. The namespace is supposed to be metadata/namespace is supposed to be completely irrelevant but is not, and the namespace transformer ends up being in a different namespace than all the other transformers. This will be an issue when we will try to apply kustomize on top of the transformer config.yaml

```go
func (kt *KustTarget) configureExternalTransformers() ([]resmap.Transformer, error) {
	ra := accumulator.MakeEmptyAccumulator()
	err := kt.accumulateResources(ra, kt.kustomization.Transformers)
	if err != nil {
		return nil, err
	}
	return kt.pLdr.LoadTransformers(kt.ldr, ra.ResMap())
}```

File: examples/configureBuiltinPlugin.md (hunk 1 above)
Comment (monopole) [reply examples/configureBuiltinPlugin.md]: Done

File: examples/configureBuiltinPlugin.md (hunk 2 above)
Comment (monopole) [reply examples/configureBuiltinPlugin.md]: Done.  Although i think we need to modify pluginator to generation this doc somehow.

I split relevant structs (e.g. configmapargs) out of types/kustomization.go into their own files so we could link to them more unambiguously from markdown.  Some structs were already in their own files, e.g image/imagetag

File: docs/plugins/builtins.md (hunk 3 above)
Comment (monopole) [reply docs/plugins/builtins.md]: done

File: docs/plugins/builtins.md (hunk 5 above)
Comment (monopole) [reply docs/plugins/builtins.md]: done

File: docs/plugins/builtins.md (hunk 7 above)
Comment (monopole) [reply docs/plugins/builtins.md]: @jbrette yes, the namespace transformer is an outlier.
see comments here: https://github.com/kubernetes-sigs/kustomize/pull/1280
it's philosophically wrong, but in practice is works and allows for a smaller config file, fwiw.
we should probably 'fix' it if for no other reason than to avoid this kind of confusion/conversation.
Not in this PR though.

File: docs/plugins/builtins.md (hunk 6 above)
Comment (monopole) [reply docs/plugins/builtins.md]: changed to "not-imporant-in-this-example" which is what i meant.
indeed, the names need to be different if using more than one.
maybe just change it to a my-image-transformer or something?

**Comment by monopole** [comment]
OK comments addressed.  sorry was doing a bunch of pushes to see how it looked since my local md renderer is busted and too lazy to fix at the moment

File: docs/plugins/builtins.md (hunk 4 above)
Comment (monopole) [reply docs/plugins/builtins.md]: done

**Comment by monopole** [comment]
I'm not happy with the section headings, e.g. 
## SecretGenerator
### Usage via kustomization.yaml
#### field name: secretGenerator

### Usage via plugin
#### Arguments
#### Example

could use some improvement, but i'd rather fix in the context of a code generator since it's so tedious to do this.

**Comment by Liujingfang1** [comment]
@monopole Thank you for this work.

/lgtm

File: docs/plugins/builtins.md (hunk 7 above)
Comment (tkellen) [reply docs/plugins/builtins.md]: Thank you for this clarification, and for this comment especially: https://github.com/kubernetes-sigs/kustomize/pull/1280#issuecomment-506904144!
